### PR TITLE
HWKAPM-764 | Percentage sampling

### PR DESCRIPTION
**Pull request by pavolloffay**
https://issues.jboss.org/browse/HWKAPM-764

This should provide the same functionality as this PR https://github.com/hawkular/hawkular-apm/pull/693

### CODE REVIEW

**Comment by pavolloffay** [comment]
@objectiser @karelhala could you please review?

**Comment by objectiser** [comment]
@pavolloffay LGTM :+1: - be good if @karelhala could comment in general on the js before merging.

**Comment by pavolloffay** [comment]
@objectiser thanks! I will wait for Karel's feedback.

File: lib/APMSpan.js
Comment (karelhala) [inline lib/APMSpan.js:394]: Honestly I don't like switch statements in Javascript, they are both memory demanding, and hard to read. There is much better way to avoid them by using [object literals](https://toddmotto.com/deprecating-the-switch-statement-for-object-literals/), so this whole switch statement can look like this:

```Javascript
if (reportingLevel) {
  const reportingTypes = {
    [REPORTING_LEVEL_NONE]: false, 
    [REPORTING_LEVEL_IGNORE]: false,
    [REPORTING_LEVEL_ALL]: true
  };
  return reportingTypes.hasOwnProperty(reportingLevel) ? reportingTypes[reportingTypes] : true;
}
```

File: lib/APMTracer.js
Comment (karelhala) [inline lib/APMTracer.js]: Might be better to indicate that you are importing whole object via `*`:
```Javascript
import * as constants from './constants';
```

File: lib/sampler.js
Comment (karelhala) [inline lib/sampler.js]: Here I'd prefer assigning function to variable:
```Javascript
class Sampler {
  isSampled(trace) {} // eslint-disable-line no-unused-vars
}

class AlwaysSample extends Sampler {
  constructor() {
    super();
    this.isSampled = alwaysSample;
  }
}

class NeverSample extends Sampler {
  constructor() {
    super();
    this.isSempled = neverSample;
  }
}
```
The benefit of this is you don't have to know what arguments will be used in `isSampled` function and in `neverSample`, `alwaysSample`. But on the other hand, it looks more ugly than returning result of function. So it's up to you what you will choose.

**Review by karelhala — CHANGES_REQUESTED** [review]
Looks good, but I'd change switch statement to Object literal and import whole object looks better with `*`.

File: lib/sampler.js
Comment (pavolloffay) [reply lib/sampler.js]: yes I like this, I used it in `PercentageSampler`

File: lib/APMSpan.js
Comment (pavolloffay) [reply lib/APMSpan.js:394]: you have a small typo there:
` return reportingTypes.hasOwnProperty(reportingLevel) ? reportingTypes[reportingLevel] : true;`

Well in some cases it's nicer to use `object literals`. I will stick with switch statement here.

**Comment by pavolloffay** [comment]
@karelhala  thanks!

**Pull request merged**
